README.md
# Svelte SSR Example with Code Split

A [Svelte 3](https://v3.svelte.technology/) server side render example, with hashed JS and CSS chunks, using [rollup](https://github.com/rollup/rollup) and [rollup-plugin-css-chunks](https://github.com/domingues/rollup-plugin-css-chunks).

## Installation

```bash
git clone https://github.com/domingues/svelte-ssr-example.git
cd svelte-ssr-example
npm install
```

## Usage

```shell script
npm run build
npm run start
```
or
```shell script
npm run start:dev
```
and open your browser
```shell script
curl http://localhost:3000/
```
you can pass props in the url
```shell script
curl http://localhost:3000/?props={%22size%22:64}
```
or submit it a POST request
```shell script
curl --header "Content-Type: application/json" \
  --request POST \
  --data '{"size":64}' \
  http://localhost:3000/
```

## Environment variables

### Build time
 - `$PUBLIC_STATIC_PATH`: path where the static files will be located,
                          local like `/static` or a remote CDN like `//abc.mycdn/project`;

### Runtime
 - `$NODE_HOST`
 - `$NODE_PORT`

## License

MIT


CHANGELOG.md
# svelte-ssr-example changelog

## 1.4.2 (07/12/2020)
 - Upgrade dependencies;

## 1.4.1 (12/11/2020)
 - Upgrade dependencies;

## 1.4.0 (08/09/2020)
 - Upgrade `rollup-plugin-extract-bundle-tree` to allow entry javascript files with hash file name;

## 1.3.6 (07/09/2020)
 - Upgrade dependencies;
 - Fix README;

## 1.3.5 (21/05/2020)
 - Upgrade dependencies;

## 1.3.4 (11/09/2019)
 - Remove unknown HTML id from script tag;
 - Upgrade dependencies;

## 1.3.3 (12/08/2019)
 - Add browser live reload to dev server;
 - Upgrade dependencies;
 
## 1.3.2 (07/08/2019)
 - Add `pre` tag server to errors;

## 1.3.1 (06/08/2019)
 - Fix `DEV_SERVER` switch;
 - Change default `NODE_HOST` to `localhost`;

## 1.3.0 (02/08/2019)
 - Allow POST requests;
 - Set HTML content type charset to UTF 8;
 - Add HTML tags to error messages;
 - Enable cross-origin resource sharing for the static files in the development server;
 - Upgrade dependencies;

## 1.2.0 (26/07/2019)

### Build
 - Update NPM scripts;
 - Rename public dist folder to "client";
 - Remove old files from dist folder;
 - Move bundle tree to root dist folder;
 - Move HTTP server to root dist folder;
 - Set public static path with environment variable:
   - `$PUBLIC_STATIC_PATH`

### HTTP Server
 - Set host and port through environment variables:
   - `$NODE_HOST`
   - `$NODE_PORT`
 - Fix import order of CSS and JS dependencies;
 - Workaround for a strange [chrome bug](https://stackoverflow.com/a/42969257);
 - Better path join;
 - Don't compress responses nor serve static files in production;
 - Set cache control for static files;
 - Add content type and length of html pages;


rollup.config.js
import multiInput from 'rollup-plugin-multi-input';
import resolve from '@rollup/plugin-node-resolve';
import svelte from 'rollup-plugin-svelte';
import css from 'rollup-plugin-css-chunks';
import bundleTree from 'rollup-plugin-extract-bundle-tree';
import importAssets from 'rollup-plugin-import-assets';
import {terser} from 'rollup-plugin-terser';
import commonjs from '@rollup/plugin-commonjs';
import replace from '@rollup/plugin-replace';
import del from 'rollup-plugin-delete';
import json from '@rollup/plugin-json';
import livereload from 'rollup-plugin-livereload';

const production = !process.env.ROLLUP_WATCH;
const publicStaticPath = process.env.PUBLIC_STATIC_PATH || '/static';

export default [
    {
        input: ['src/*.svelte'],
        output: {
            dir: 'dist/client',
            sourcemap: !production,
            format: 'esm',
            chunkFileNames: '[hash].js',
            entryFileNames: '[hash].js',
        },
        plugins: [
            del({targets: 'dist/client/*'}),
            multiInput(),
            resolve({browser: true}),
            svelte({
                compilerOptions: {
                    dev: !production,
                    immutable: true,
                    hydratable: true,
                },
            }),
            css({
                chunkFileNames: '[hash].css',
                entryFileNames: '[hash].css',
                sourcemap: !production,
            }),
            importAssets({
                fileNames: '[hash].[ext]',
                publicPath: publicStaticPath,
            }),
            bundleTree({
                file: 'dist/client-tree.json',
            }),
            !production && livereload({
                watch: 'dist/server',
                delay: 750,
            }),
            production && terser(),
        ],
        watch: {
            clearScreen: false,
        },
    },
    {
        input: ['src/*.svelte'],
        output: {
            dir: 'dist/server',
            format: 'cjs',
            exports: 'default',
        },
        plugins: [
            del({targets: 'dist/server/*'}),
            multiInput(),
            resolve(),
            svelte({
                compilerOptions: {
                    dev: !production,
                    immutable: true,
                    hydratable: true,
                    generate: 'ssr',
                },
            }),
            css({
                emitFiles: false,
            }),
            importAssets({
                fileNames: '[hash].[ext]',
                publicPath: publicStaticPath,
                emitAssets: false,
            }),
            production && terser(),
        ],
    },
    {
        input: 'src/server.js',
        output: {
            file: 'dist/server.js',
            format: 'cjs',
        },
        plugins: [
            replace({
                STATIC_PATH: publicStaticPath,
                DEV_SERVER: !production,
            }),
            resolve({preferBuiltins: true}),
            commonjs(),
            json(),
            production && terser(),
        ],
    },
];


src/server.js
import fs from 'fs';
import path from 'path';
import urljoin from 'url-join';
import polka from 'polka';
import sirv from 'sirv';
import compression from 'compression';
import bodyParser from 'body-parser';

function parseBundleTree(file, public_static_path) {
    const bundle_tree = JSON.parse(fs.readFileSync(path.join(__dirname, file)));

    const dfs_cache = {};

    function dfs(file) {
        const file_public_path = urljoin(public_static_path, file);
        if (!dfs_cache[file_public_path]) {
            const css = [], js = [];
            if (bundle_tree[file].imports) {
                for (const f of bundle_tree[file].imports) {
                    const f_public_path = urljoin(public_static_path, f);
                    if (!bundle_tree[f].isAsset) {
                        const { css: c, js: j } = dfs(f);
                        css.push(...c);
                        js.push(...j);
                        js.push(f_public_path);
                    } else if (bundle_tree[f].isAsset && f.endsWith('.css')) {
                        css.push(f_public_path);
                    }
                }
            }
            dfs_cache[file_public_path] = { css, js };
        }
        return dfs_cache[file_public_path];
    }

    const clientImports = {};
    const serverClientMap = {};
    for (const file in bundle_tree) {
        if (bundle_tree[file].isEntry !== true) continue;
        const { css, js } = dfs(file);
        const file_public_path = urljoin(public_static_path, file);
        clientImports[file_public_path] = { css: [...new Set(css)], js: [...new Set(js)] };
        serverClientMap[bundle_tree[file].chunkName] = file_public_path;
    }

    return { clientImports, serverClientMap };
}

function render(component_path, props_json, head, html) {
    // the "<script> </script>" (with space) is to prevent a chrome bug https://stackoverflow.com/a/42969257
    return `<!doctype html>
<html>
<head>
	<meta charset='utf8'>
	<meta name='viewport' content='width=device-width, initial-scale=1'>
	${head}
</head>
<body>
	<div id='content'>${html}</div>
	<script type='module'>
		import App from '${component_path}';
		new App({
			target: document.getElementById('content'),
			hydrate: true,
			props: ${props_json}
		});
	</script>
	<script> </script>
</body>
</html>`;
}

function buildPage(component, props_json, clientImports, serverClientMap) {
    let Component, props, head, html;
    try {
        Component = require(path.join(__dirname, 'server', component));
    } catch (err) {
        return { error: `importing module "${component}"`, code: err };
    }
    try {
        props = JSON.parse(props_json);
    } catch (err) {
        return { error: 'parsing props JSON', code: err };
    }
    try {
        const component = Component.render(props);
        head = component.head;
        html = component.html;
    } catch (err) {
        return { error: 'rendering module', code: err };
    }

    const component_path = serverClientMap[component];
    clientImports[component_path].css.forEach(f =>
        head += `\n\t<link rel='stylesheet' href='${f}'>`);
    clientImports[component_path].js.forEach(f =>
        head += `\n\t<script type='module' src='${f}'></script>`);
    head += `\n\t<script type='module' src='${component_path}'></script>`;
    const page = render(component_path, props_json, head, html);

    return { page };
}

const HOST = process.env.NODE_HOST || 'localhost';
const PORT = process.env.NODE_PORT || 3000;
const PUBLIC_STATIC_PATH = 'STATIC_PATH';
const DEV = 'DEV_SERVER' === 'true';

const { clientImports, serverClientMap } = parseBundleTree('client-tree.json', PUBLIC_STATIC_PATH);

const server = polka();
if (DEV) {
    server.use(compression({ threshold: 0 }));
    let p = PUBLIC_STATIC_PATH.replace(/^(^\w+:|^)\/\/[\w.:]+/, '').replace(/[/]*$/, '');
    server.use(p, sirv(path.join(__dirname, 'client'), {
        setHeaders: res => {
            res.setHeader('Access-Control-Allow-Origin', '*');
        },
        maxAge: 31536000,
        immutable: true
    }));
}
const respond = (req, res) => {
    let code, html;
    try {
        const props = req.method === 'GET' ? (req.query.props ? req.query.props : '{}') : (req.body ? req.body : '{}');
        const result = buildPage(req.params['*'] ? req.params['*'] : 'index', props, clientImports, serverClientMap);
        if (result.error) {
            code = 500;
            html = `<!DOCTYPE html><html lang="en"><body>Error ${result.error}:<br/><pre>${result.code}</pre></body></html>`;
        } else {
            code = 200;
            html = result.page;
        }
    } catch (err) {
        console.error(err); // eslint-disable-line
        code = 500;
        html = `<!DOCTYPE html><html lang="en"><body>Unknown error:<br/><pre>${err}</pre></body></html>`;
    }
    res.writeHead(code, {
        'Content-Type': 'text/html; charset=utf-8',
        'Content-Length': html.length,
    });
    res.end(html);
};
server.use(bodyParser.text({ type: '*/json' }));
server.get('/*', respond)
    .post('/*', respond)
    .listen(PORT, HOST, err => {
        if (err) throw err;
        console.log(`Running on http://${HOST}:${PORT}`); // eslint-disable-line
    });
